### PR TITLE
Mesh Render Extent Settings

### PR DESCRIPTION
**Pull request by JanCaha**
## Description

This PR brings to renderer for Mesh Contours option to set renderer Min Max based on extent (specific canvas, changing canvas) in the same way as renderers for Raster Layer. 

Besides that `minimumMaximumActiveScalarDataset( const QgsRectangle &extent, const QgsMeshDatasetIndex &datasetIndex, double &min, double &max )` is added to MeshLayer to obtain Min Max from specific Scalar Dataset Index for  given extent.

### CODE REVIEW

File: src/core/mesh/qgsmeshrenderersettings.h [hunk 1]
```diff
@@ -115,6 +114,17 @@ class CORE_EXPORT QgsMeshRendererScalarSettings
...
 
+    /**
+     *  This enumerator describes the extent used to compute min/max values
+     */
+    enum class MinMaxValueType
+    {
+      UserDefined, //User defined Min Max values
+      WholeMesh, // Values from whole mesh
+      FixedCanvas, //Fixed Min Max Values set from canvas
+      InteractiveFromCanvas, //Constantly updated from extent of the canvas is used to compute statistics
```
Comment (DelazJ) [inline src/core/mesh/qgsmeshrenderersettings.h:125]: Is there something missing in this sentence?

File: src/core/mesh/qgsmeshrenderersettings.h [hunk 2]
```diff
@@ -176,12 +186,28 @@ class CORE_EXPORT QgsMeshRendererScalarSettings
      */
     void setEdgeStrokeWidthUnit( Qgis::RenderUnit edgeStrokeWidthUnit );
 
+    /**
+     * Gets the extent type for minimum maximum calculation
```
Comment (DelazJ) [inline src/core/mesh/qgsmeshrenderersettings.h:190]:
```suggestion
     * Gets the extent type for minimum and maximum calculation
```
?

File: src/core/mesh/qgsmeshrenderersettings.h [hunk 3]
```diff
@@ -176,12 +186,28 @@ class CORE_EXPORT QgsMeshRendererScalarSettings
...
+     *
+     * \since QGIS 3.42
+     */
+    QgsMeshRendererScalarSettings::MinMaxValueType minMaxValueType() const { return mMinMaxValueType; }
+
+    /**
+     * Sets the extent type for minimum maximum calculation
```
Comment (DelazJ) [inline src/core/mesh/qgsmeshrenderersettings.h:197]:
```suggestion
     * Sets the extent type for minimum and maximum calculation
```
?

File: tests/src/core/testqgsmeshlayer.cpp [hunk 4]
```diff
@@ -2366,5 +2368,108 @@ void TestQgsMeshLayer::testHaveSameParentQuantity()
...
+  QCOMPARE( min, 1 );
+  QCOMPARE( max, 2 );
+
+  extent = QgsRectangle( 900, 1900, 1100, 2100 );
+  found = mMdalLayer->minimumMaximumActiveScalarDataset( extent, datasetIndex, min, max );
+  QVERIFY( found );
+  QCOMPARE( min, 1 );
+  QCOMPARE( max, 1 );
+
+  // cannot get data for vector dataset - index points to wrong dadataset index
```
Comment (DelazJ) [inline tests/src/core/testqgsmeshlayer.cpp:2412]:
```suggestion
  // cannot get data for vector dataset - index points to wrong dataset index
```

File: src/core/mesh/qgsmeshrenderersettings.h [hunk 5]
```diff
@@ -115,6 +114,17 @@ class CORE_EXPORT QgsMeshRendererScalarSettings
...
     };
 
+    /**
+     *  This enumerator describes the extent used to compute min/max values
+     */
+    enum class MinMaxValueType
+    {
+      UserDefined, //User defined Min Max values
+      WholeMesh, // Values from whole mesh
+      FixedCanvas, //Fixed Min Max Values set from canvas
```
Comment (DelazJ) [inline src/core/mesh/qgsmeshrenderersettings.h:124]: Should these be?
```suggestion
      UserDefined, //!< User defined Min Max values
      WholeMesh, //!< Values from whole mesh
      FixedCanvas, //!< Fixed Min Max Values set from canvas
```

**Review by DelazJ — COMMENTED** [review]
@JanCaha Can you update your top message with a screenshot for the docs and changelog, please? Thanks.

File: src/core/mesh/qgsmeshrenderersettings.h (hunk 5 above)
Comment (JanCaha) [reply src/core/mesh/qgsmeshrenderersettings.h:124]: fix texts

**Comment by nyalldawson** [comment]
@JanCaha have you seen https://github.com/qgis/QGIS/pull/59444 ? The current way the raster layer handles extent based rendering is fragile, and is being reworked in that PR. Are you able to adapt to the same approach as is used there?

**Comment by JanCaha** [comment]
@nyalldawson I did not see that one. Actually this PR is older then the raster one :-) 

It should not be a problem to adapt to that PR. Let me wait for #59444 to be merged and then rework this.

**Pull request closed without merging**